### DISCUSSION TITLE
Support for Both Swift 5 and 6 in Package Builds

### DISCUSSION THREAD

**Discussion by onevcat** [SwiftPackageIndex/SwiftPackageIndex-Server#3138] [Q&A] [answered]
Hi,

I've updated my package to be Swift 6 compatible and adjusted the `SWIFT_VERSION` in my scheme to 6.0. I've also ensured that it remains capable of building under Swift 5. Despite these changes, I'm encountering a build error on the Swift Package Index (SPI) at [this link](https://swiftpackageindex.com/builds/43A9ED3F-D5D3-4C01-9CF0-5D51A0AFF745).

Obviously, the old Xcode does not contain the new Swift version specified in the project setting. I believe specifying `SWIFT_VERSION=5.0` via xargs might resolve the issue, but I'm unsure how to implement this and pass the build argument to SPI.

Could you advise on how to configure the package to support both Swift 5 and 6 seamlessly, or suggest any tools that might help?

Thank you for your assistance.

**Comment by finestructure** [comment]
I believe you'd either have to add a `Package@swift-5.10.swift` for your Swift 5 config or a `Package@swift-6.0.swift` for Swift 6. That way you can configure your build depending on the Swift version your package is built with.

**Comment by finestructure** [comment]
I guess you could also try `#ifdef compiler(>=6.0)` (or something like this, I've not checked the exact syntax) within a single file to manage adding a Swift language version.

**Comment by onevcat** [comment]
@finestructure 

Thank you for your response. I will attempt to specify a version in the `Package.swift` and observe the outcome.

However, as detailed in the [build results](https://swiftpackageindex.com/builds/43A9ED3F-D5D3-4C01-9CF0-5D51A0AFF745), the command utilized for the build appears to be:

```
env DEVELOPER_DIR=/Applications/Xcode-15.4.0.app xcrun xcodebuild ... build -scheme Kingfisher -destination generic/platform=iOS
```

This command employs the project scheme and settings defined within the Xcode project file to build the target. Given the error message "SWIFT_VERSION '6.0' is unsupported, supported versions are: 4.0, 4.2, 5.0. (in target 'Kingfisher' from project 'Kingfisher')", it seems to be a project settings issue.

Could you please clarify how adjustments to the package file might influence this situation (If I am correct, if built with swift package manager, the project setting won't be used at all)?

> I guess you could also try #ifdef compiler(>=6.0) (or something like this, I've not checked the exact syntax) within a single file to manage adding a Swift language version.

And thanks for the comments. It is `#if swift(>=6)`, and I have already used it to make my project build for both Swift 5 and 6. Currently, I need a way to pass `SWIFT_VERSION=5.0` to the build command...

**Reply by finestructure** [reply, marked as answer]
I believe you should be able to set `SWIFT_VERSION` to `5` via the scheme. If that's problematic for your current scheme, you could add another one and reference it via the `scheme:` in `.spi.yml`.

Maybe this page here has some useful info? https://www.swift.org/migration/documentation/swift-6-concurrency-migration-guide/swift6mode/

**Reply by onevcat** [reply]
Thank you for your response. I’ve decided to revert to `SWIFT_VERSION=5.0` for the time being, as upgrading to the latest Swift version hasn't provided significant benefits to my project. I appreciate your patient support!
